[nicholasohjj] iP

nicholasohjj opened the pull request:
# iP: GitHub Flavored Markdown Update

This PR includes an update to the description with various GitHub Flavored Markdown elements.

## Features Included:
- A heading ✅
- A bullet list ✅
- A numbered list ✅
- A fenced code block with syntax highlighting ✅
- A task list ✅
- An emoji ✅
- A blockquote ✅
- A hyperlink ✅
- Inline code ✅
- Text formatting (bold, italic, strikethrough) ✅

### Steps Completed:
1. Created the initial PR
2. Edited the PR description
3. Applied GitHub Flavored Markdown elements

### Code Example:
```java
public class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello, GitHub Flavored Markdown!");
    }
}
```
### Task List:

- [x]  Understand GFMD syntax
- [x]  Edit the PR description
- [ ]  Get approval from reviewer

> "Markdown is a lightweight and easy-to-use syntax for styling all forms of writing on GitHub." – GitHub Docs
> 

### Additional Formatting:

- **Bold text**
- *Italic text*
- ~~Strikethrough text~~
- `Inline code example`

🔗 [GitHub Flavored Markdown Guide](https://guides.github.com/features/mastering-markdown/) 🚀



### **What Each Element Does**
- `#` creates a **heading**.
- `-` creates a **bullet list**.
- `1.` creates a **numbered list**.
- Triple backticks with a language (e.g., `java`) create a **fenced code block with syntax highlighting**.
- `- [ ]` and `- [x]` create a **task list** (unchecked and checked, respectively).
- `> ` creates a **blockquote**.
- `:rocket:` (or other emoji codes) insert an **emoji**.
- `[text](URL)` creates a **hyperlink**.
- Backticks (`code`) format **inline code**.
- `**bold**`, `*italic*`, `~~strikethrough~~` apply **text formatting**.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

nicholasohjj committed:
Rename Duke.java to NiniNana.java, modify chat greeting and goodbye

nicholasohjj committed:
Modiy chat to echo command

nicholasohjj committed:
Add storing functionality in chat

nicholasohjj committed:
Add storing functionality in chat

nicholasohjj committed:
Refactor and add Task class. Add functionality to mark as done

nicholasohjj committed:
Create subclasses ToDo, Event, Deadline Tasks. Add TaskCommands for creating subclasses

nicholasohjj committed:
Edit runtest.sh and runtest.bat to use NiniNana.java

nicholasohjj committed:
Add exception handling to chat

nicholasohjj committed:
Add delete support for chat

nicholasohjj committed:
Refactor code with enums

nicholasohjj committed:
Refactor exceptions into exceptions folder

nicholasohjj committed:
Add serialize and deserialize methods for Task class and subclasses. Add File reading and writing functionality in NiniNana.java

nicholasohjj committed:
Merge branch 'branch-Level-7'

nicholasohjj committed:
amend gitignore

nicholasohjj committed:
Merge branch 'branch-Level-7'

nicholasohjj committed:
Refactor date attributes to use LocalDateTime format

nicholasohjj committed:
Merge branch-Level-8 to master

nicholasohjj committed:
Refactor code with more OOP and Command pattern

nicholasohjj committed:
merge branch-A-MoreOOP to master

nicholasohjj committed:
Add message content to exceptions

nicholasohjj committed:
Merge add-gradle-support to master

nicholasohjj committed:
Merge master to branch-A-Gradle

nicholasohjj committed:
Set up gradle

nicholasohjj committed:
Merge branch-A-Gradle to master

nicholasohjj committed:
Add junit tests for components and tasks

nicholasohjj committed:
Add junit tests for components and tasks

nicholasohjj committed:
Create jar file (ignored)

nicholasohjj committed:
Add javadocs to all classes/methods

nicholasohjj committed:
Merge branch-A-JavaDoc to master

nicholasohjj committed:
Edit classes to follow code standard

nicholasohjj committed:
Fix merge conflicts with branch-A-JavaDocs

nicholasohjj committed:
Add functionality to find tasks by keyword

nicholasohjj committed:
Resolve merge conflicts with branch-Level-9

nicholasohjj committed:
Edit ParserTest

nicholasohjj committed:
Remove tatus

nicholasohjj committed:
Fix code style with checkstyle

nicholasohjj committed:
merge branch-A-CheckStyle to master

nicholasohjj committed:
Merge master to branch-Level-10

nicholasohjj committed:
Remove bug from ToDoTaskTest

nicholasohjj committed:
Merge master to branch-Level-10

nicholasohjj committed:
Remove bug from ToDoTaskTest

nicholasohjj committed:
Merge master to branch-Level-10

nicholasohjj committed:
Refactor codebase to use javafx and fxml

nicholasohjj committed:
merge branch-level-10 to master

nicholasohjj committed:
Refactor unmark, mark, delete commands to accept varargs indices

nicholasohjj committed:
merge branch-A-Varargs

weien02 commented on src/main/java/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @param taskIndices The indices of the tasks to be deleted (zero-based).
+     */
+    public DeleteCommand(int... taskIndices) {
```
I like your natural implementation of Varargs!

nicholasohjj replied:
Thanks!

weien02 commented on src/main/java/components/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+        });
+    }
+
+    private LocalDateTime getTaskDate(Task task) {
+        if (task instanceof DeadlineTask) {
+            return ((DeadlineTask) task).getDeadline();
+        } else if (task instanceof EventTask) {
+            return ((EventTask) task).getStartDateTime();
+        }
+        return null;
+    }
```
Would it be better to encapsulate this within the respective Task subclasses? That could also make better use of polymorphism.

nicholasohjj replied:
fixed in latest commit

weien02 commented on src/main/java/commands/FindCommand.java:
```diff
@@ -0,0 +1,52 @@
+     */
+    @Override
+    public String execute(TaskList taskList, Ui ui, Storage storage) throws NiniException {
+        List<Task> matchingTasks = taskList.getTasks().stream()
+                .filter(task -> task.getDescription().toLowerCase().contains(keyword.toLowerCase()))
+                .collect(Collectors.toList());
```
Would it be better to encapsulate this step into the TaskList class instead?

nicholasohjj replied:
fixed in latest commit

weien02 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,158 @@
+     * @throws NiniException If the data is invalid, incorrectly formatted, or incomplete.
+     */
+    public static Task deserialize(String data) throws NiniException {
+        if (data == null || data.isEmpty()) {
+            throw new InvalidDataException("Cannot deserialize null or empty data.");
+        }
+        String[] parts = data.split("\\|");
+
+        if (parts.length < 3) {
+            throw new InvalidDataException("Incomplete data for task deserialization.");
+        }
+
+        if (!parts[DONE_INDEX].equals("1") && !parts[DONE_INDEX].equals("0")) {
+            throw new InvalidDataException("Data has invalid values for is_done status");
+        }
+
+        String type = parts[TYPE_INDEX].trim();
+        boolean isDone = parts[DONE_INDEX].trim().equals("1");
+        String description = parts[DESCRIPTION_INDEX].trim();
+
+        switch (type) {
+        case "T":
+            return new ToDoTask(description, isDone);
+        case "D":
+            if (parts.length < 4) {
+                throw new InvalidDataException("Missing deadline information.");
+            }
+            String deadlineStr = parts[3].trim();
+            return new DeadlineTask(description, deadlineStr, isDone);
+        case "E":
+            if (parts.length < 5) {
+                throw new InvalidDataException("Missing event start or end time.");
+            }
+            String fromStr = parts[3].trim();
+            String toStr = parts[4].trim();
+            return new EventTask(description, fromStr, toStr, isDone);
+        default:
+            throw new InvalidDataException("Unknown task type.");
+        }
+    }
```
Interesting implementation, but would like to seek an understanding of why you chose to implement this in the Task class, instead of let's say Storage or Parser?

nicholasohjj replied:
The `deserialize` method is responsible for reconstructing a `Task` object from a string. Since each task type (ToDoTask, DeadlineTask, EventTask) has unique properties (e.g., deadlines, start/end times), it makes sense for `Task` to own this logic

weien02 commented on src/main/java/components/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+     */
+    public Task getTask(int index) {
+        if (!isValidIndex(index)) {
+            throw new IndexOutOfBoundsException("Error: Invalid task index.");
```
Would an implementation of a separate exception extending NiniException help you finetune your error handling?

weien02 reviewed (commented):
Nice work! Do remember to add JavaDocs for your functions.

nicholasohjj committed:
Encapsulate task finding into TaskList class

alexandtheoh commented on src/main/java/NiniNana.java:
```diff
@@ -6,13 +11,26 @@
+    private static final String FILENAME = "./data/chat.txt";
+
+    private static void ensureFileDirectoryExists() {
+        File directory = new File("./data");
```
Interesting, perhaps u can derive the directory from the file name instead of hardocding it.

nicholasohjj replied:
resolved

alexandtheoh reviewed (commented):
very nice work

nicholasohjj committed:
Use assert feature to document important assumptions

Certain assumptions should always hold at various points in the code, but they are not explicitly enforced. Without proper documentation, these assumptions may be overlooked or unintentionally broken.

Using the assert feature allows these assumptions to be explicitly stated, making the code more self-explanatory and helping with debugging by detecting violations early.

Let's introduce assertions to document critical assumptions throughout the codebase, ensuring that expected conditions are met while keeping the implementation clear.

nicholasohjj committed:
Merge pull request #1 from nicholasohjj/branch-A-Assertions

Use assert feature to document important assumptions

nicholasohjj committed:
Refactor all classes to follow code quality standards

Classes in components and commands folder have several issues affecting code quality and maintainability

- Parser: parseEvent() throws ArrayIndexOutOfBoundsException instead
  of InvalidFormatException when input format is incorrect.
- MainWindow: The bot image was incorrectly loading the user image, and
  UI updates were not thread-safe.
- DialogBox: FXMLLoader failures were silently ignored, leading to
  potential UI crashes.

Let's,
- Fix parseEvent() to correctly validate input format before splitting
- Ensure all InvalidFormatException cases are properly thrown
- Correct MainWindow to load the correct bot image and prevent UI
  crashes by using Platform.runLater()
- Improve DialogBox error handling by throwing a RuntimeException
  instead of just printing stack traces
- Implement Objects.requireNonNull() checks for better null safety

This improves code robustness and ensures better debugging when issues
occur.

nicholasohjj committed:
Merge pull request #2 from nicholasohjj/branch-A-CodeQuality

Refactor all classes to follow code quality standards

nicholasohjj committed:
Refactor to use streams API

nicholasohjj committed:
Merge pull request #3 from nicholasohjj/branch-A-Streams

Refactor to use streams API

nicholasohjj committed:
Add CI yml file to github actions

nicholasohjj committed:
Merge pull request #4 from nicholasohjj/branch-A-CI

Add CI yml file to github actions

nicholasohjj committed:
Modify CI yml file to reduce redundant steps

nicholasohjj committed:
Modify dateTimeParser to use Locale.ENGLISH

nicholasohjj committed:
Modify EventTask and DeadlineTask to meet code checks

nicholasohjj committed:
Remove io redirection tests on gradle.yml

nicholasohjj committed:
Merge pull request #5 from nicholasohjj/branch-A-CI

Branch a ci

nicholasohjj committed:
Edit Parser.java to remove redundant parameters

nicholasohjj committed:
Add Contact, ContactList, ContactStorage and remove Ui class

nicholasohjj committed:
Implement all commands for contacts

nicholasohjj committed:
Merge pull request #6 from nicholasohjj/branch-D-Contacts

Branch d contacts

nicholasohjj committed:
Improve the GUI to make it more polished

nicholasohjj committed:
Merge branch to master

nicholasohjj committed:
Give a unique personality

nicholasohjj committed:
Add a User Guide

nicholasohjj committed:
Update README.md

nicholasohjj committed:
Modify readme

nicholasohjj committed:
Modify readme

nicholasohjj committed:
Modify UG
